[RFC] linux: translate config to structured config

teto opened the pull request:
Follow up of https://github.com/NixOS/nixpkgs/pull/12158 @copumpkin . I am looking for comments before polishing as compiling kernels is not especially fast here.

The set is then converted into a string so backwards compatibility kinda applies (users can still specify their kernel extraConfig as a string) but it should be deprecated in favor of  sets.

###### Motivation for this change

- I want to be able to convert "module" answers into "yes" and vice-versa
- with the current kernel config system, the first defined is taken into account so
```
extraConfig = ''
OPENVSWITCH y 
OPENVSWITCH m
'';
```
will end up with `yes`. Even though this PR doesn't solve this problem, it paves the way for some kind of mkMerge mechanism.


###### Things done
-  added structuredExtraConfig and mkValuePreprocess to buildLinux args, I kept `extraConfig` as a string for backawards compatibility, structuredExtraConfig must be a flat set e.g. `{ BPF = yes; BPF_SYSCALL = y; }`
- `mkValuePreprocess` allows to override each setting, aka convert 'yes' into 'modules'  or  vice-versa


I test the changes with this overlay expression and tried compiling a few kernels (can't do all though with my poor machine).
```
  linux_test = prev.linux_4_16.override {
    ignoreConfigErrors=true;
    autoModules = false;
    kernelPreferBuiltin = true;
  };
```

Future work
[ ] Convert patches to this new format
[ ] make packages describe their kernel requirements (bcc needs BPF, openvswitch needs OPENVSWITCH etc)
For instance the openvswitch package could contain hints to the configuration:
```
 with lib.kernel;
   passthru.kernelExtraConfig = with callPackage ./lib/kernel.nix { version = kernel.version; }; {
     OPENVSWITCH  = whenAtLeast "2.0" yes;
   };
```
and have the module actually enforce that configuration:
```
    boot.kernelPatches = [ {
         name = "openvswitch";
         patch = null;
         extraConfig = lib.kernel.generateKConf pkgs.openvswitch.passthru.kernelExtraConfig;
       }
     ];
```


<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

copumpkin committed:
linux: translate config to structured config

Instead of using a string to describe kernel config, use a nix
attribute set, then converted to a string.
- allows to override the config, aka convert 'yes' into 'modules' or
vice-versa
- while for now merging different configs is still crude (last spec wins),
at least there should be only one CONFIG_XYZ value compared to the current string
config where the first defined would be used and others ignored.

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.41393&attempt_id=1b175408-f0ae-4407-bfa8-83094012a073)

Attempted: linux

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/si1gisyb8ly0vnrhgdx5qxdzngyvsi5v-linux-4.14.47
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.41393&attempt_id=8f7a1f9d-2ea5-49ca-8606-7b7d7882b19d)

Attempted: linux

<details><summary>Partial log (click to expand)</summary><p>

```
these paths will be fetched (48.28 MiB download, 78.80 MiB unpacked):
  /nix/store/swwnj5b8jfz1bi6fvvfvlcgqwwsbkzni-linux-4.14.47
copying path '/nix/store/swwnj5b8jfz1bi6fvvfvlcgqwwsbkzni-linux-4.14.47' from 'https://cache.nixos.org'...
warning: SQLite database '/nix/var/nix/db/db.sqlite' is busy
/nix/store/swwnj5b8jfz1bi6fvvfvlcgqwwsbkzni-linux-4.14.47
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.41393&attempt_id=68406165-b113-4c64-937d-60c95048ba84)

Attempted: linux

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/nn55hl7zxl8cx9xl0bv7ra8jz2bqjczr-linux-4.14.49
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.41393&attempt_id=a43c8d8b-2e87-4626-b73e-f22fd1e4165c)

Attempted: linux

<details><summary>Partial log (click to expand)</summary><p>

```
these paths will be fetched (48.27 MiB download, 78.79 MiB unpacked):
  /nix/store/6wpbdldldwn76lxq1m3vjrgp9a15aqxz-linux-4.14.49
copying path '/nix/store/6wpbdldldwn76lxq1m3vjrgp9a15aqxz-linux-4.14.49' from 'https://cache.nixos.org'...
/nix/store/6wpbdldldwn76lxq1m3vjrgp9a15aqxz-linux-4.14.49
```
</p></details>

teto commented:
I updated the default config so that it reflects exaclty (notwithstanding my mistakes) the current common-config.nix.
When I try to run `nox-review wip --against=upstream/master` I get this kind of error http://nixpaste.lbr.uno/-fsL_q4k?nix   so I accept any tip on how to tackle this kind of "infinite recursion" problem

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.41393&attempt_id=27c19708-2594-4d5e-b711-2e233ff9c6f5)

Attempted: linux

<details><summary>Partial log (click to expand)</summary><p>

```
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
/nix/store/cv3cfq0axhlcqg0khw1lrxddxsg03nyb-linux-4.14.50
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.41393&attempt_id=10d477df-feff-43a8-93b5-450851cf9b52)

Attempted: linux

<details><summary>Partial log (click to expand)</summary><p>

```
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
cannot find section .dynamic
/nix/store/bja0q8gf2bxv798kxmqhjw3sqzlhxadv-linux-4.14.50
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.41393&attempt_id=7a9fd5bb-882c-4fcf-9026-0b85360c582d)

Attempted: linux

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/cv3cfq0axhlcqg0khw1lrxddxsg03nyb-linux-4.14.50
```
</p></details>

teto commented:
I've updated common-config to reflect all the changes since copumpkin PRs and i could compile 4.17. It should be ready for review (apart from the messy history)

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.41393&attempt_id=5e9da905-aacb-40ad-a459-ff7dc3690228)

Attempted: linux

<details><summary>Partial log (click to expand)</summary><p>

```
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
cannot find section .dynamic
/nix/store/bja0q8gf2bxv798kxmqhjw3sqzlhxadv-linux-4.14.50
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.41393&attempt_id=e31acab9-f905-4c23-849c-8c9ea95730dc)

Attempted: linux

<details><summary>Partial log (click to expand)</summary><p>

```
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
cannot find section .dynamic
/nix/store/bja0q8gf2bxv798kxmqhjw3sqzlhxadv-linux-4.14.50
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.41393&attempt_id=8f9e50e3-b274-4b3b-9df6-1efa29dd7a5f)

Attempted: linux

<details><summary>Partial log (click to expand)</summary><p>

```
  CC [M]  drivers/staging/speakup/fakekey.o
  CC [M]  drivers/staging/speakup/main.o
  CC [M]  drivers/staging/rtl8723bs/os_dep/ioctl_cfg80211.o
  CC [M]  drivers/staging/rtlwifi/halmac/halmac_88xx/halmac_api_88xx_usb.o
  CC [M]  drivers/staging/rtlwifi/halmac/halmac_88xx/halmac_api_88xx_sdio.o
  CC [M]  drivers/staging/rtl8723bs/os_dep/ioctl_linux.o
  CC [M]  drivers/staging/speakup/keyhelp.o
  CC [M]  drivers/staging/speakup/kobjects.o
building of '/nix/store/24plzvpzg33v84pmfd7miq69imcyd8ff-linux-4.14.50.drv' timed out after 1800 seconds
error: build of '/nix/store/24plzvpzg33v84pmfd7miq69imcyd8ff-linux-4.14.50.drv' failed
```
</p></details>

NeQuissimus commented on pkgs/os-specific/linux/kernel/common-config.nix:
Since there are no kernels <4.4, should we maybe just remove all the options for really old kernels?
And things such as `whenAtLeast "3.4" yes` may as well just be `yes`

teto replied:
Now that it's there I would say keep it, some people might run their own 3.X kernel and might appreciate it ? I am good either way.

NeQuissimus replied:
Maybe we will do that in a future PR. I highly doubt somebody uses `master` / future-18.09 with a 3.x kernel :D

NeQuissimus commented on pkgs/os-specific/linux/kernel/common-config.nix:
We may want to add the other governors as options here

dezgeg replied:
Let's do such changes separately from the structured-options refactoring; that way it is much easier to verify there weren't unintended changes.

teto replied:
agreed, I had to remove some copumpkin CONFIG_ additions that would generate errors on recent kernels.

NeQuissimus reviewed (commented):
What happens to dependencies? If I set module A to "n" but module B to "y" despite B depending on A, the kernel config will error out, correct?
Is there something simple we can do to detect this? (I can't see a simple way)

teto commented:
> What happens to dependencies? If I set module A to "n" but module B to "y" despite B depending on A, > the kernel config will error out, correct?
> Is there something simple we can do to detect this? (I can't see a simple way)
This is the current behavior and the PR doesn't change that.
I discovered yesterday https://github.com/ulfalizer/Kconfiglib which might help in that regard at a later date, I am not sure. The kernel has also a script to merge different configs https://github.com/torvalds/linux/blob/master/scripts/kconfig/merge_config.sh, I wonder if this could be used to fix some of our dependencies problem. But I would rather leave this exploration to future PRs.

dezgeg commented:
I noticed some minor differences in the configs, feel free to squash the fixes from https://github.com/NixOS/nixpkgs/compare/master...dezgeg:structured-kernel-conf-fixes?expand=1

teto commented:
wow thanks that's super nice, you caught quite a few ones. I squashed and rebased. I added a commit to remove references to any kernel < 4.4 so that it can be reverted later.
I was able to compile the config for 4.4 and 4.16.

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.41393&attempt_id=aea6d85d-f0eb-4898-8ff3-4f14f3ae72ed)

Attempted: linux

<details><summary>Partial log (click to expand)</summary><p>

```
  CC [M]  drivers/staging/rtl8192e/rtl8192e/rtl_pm.o
  CC [M]  drivers/video/fbdev/sstfb.o
  CC [M]  drivers/staging/rtl8712/usb_halinit.o
  CC [M]  drivers/staging/rtl8192u/r8190_rtl8256.o
  CC [M]  drivers/staging/rtl8192e/rtl8192e/rtl_ps.o
  CC [M]  drivers/video/fbdev/cirrusfb.o
  CC [M]  drivers/staging/rtl8712/usb_ops.o
  CC [M]  drivers/staging/rtl8192u/r819xU_phy.o
building of '/nix/store/kvi48kp4zzpbaqyvv313kpjygxwc2sra-linux-4.14.51.drv' timed out after 1800 seconds
error: build of '/nix/store/kvi48kp4zzpbaqyvv313kpjygxwc2sra-linux-4.14.51.drv' failed
```
</p></details>

teto committed:
kernel: remove conditions based on 3.X kernels

The oldest kernel in nixpkgs being 4.4, we get rid of checks for older
kernels.

dezgeg commented on pkgs/os-specific/linux/kernel/common-config.nix:
Hmm, what's this for?

teto replied:
when reworking the original PR, the features system had changed a bit and I came up with the hack to remove later. I had forgotten about it; I pushed a 3rd commit which uses features.xdom0 instead. It should fix it.

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.41393&attempt_id=40c86f0c-c47b-42e5-af53-f11af39a64b4)

Attempted: linux

<details><summary>Partial log (click to expand)</summary><p>

```
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
cannot find section .dynamic
/nix/store/qp98m3d4xa1f7x8xbnvwa7wczfbllav7-linux-4.14.51
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.41393&attempt_id=e7f78eee-3105-421c-94d6-687c258dce97)

Attempted: linux

<details><summary>Partial log (click to expand)</summary><p>

```
  CC [M]  drivers/usb/storage/isd200.o
  CC [M]  drivers/usb/storage/jumpshot.o
  CC [M]  drivers/usb/storage/karma.o
  CC [M]  drivers/usb/storage/onetouch.o
  CC [M]  drivers/usb/storage/realtek_cr.o
  CC [M]  drivers/usb/storage/sddr09.o
  CC [M]  drivers/usb/storage/sddr55.o
  CC [M]  drivers/usb/storage/shuttle_usbat.o
building of '/nix/store/qq0jx495ka6g8s75h3zn3nhby0gc3nbv-linux-4.14.51.drv' timed out after 1800 seconds
error: build of '/nix/store/qq0jx495ka6g8s75h3zn3nhby0gc3nbv-linux-4.14.51.drv' failed
```
</p></details>

teto committed:
linux: fix features system

dezgeg commented:
I verified with scripts that after a minor fix I squashed in there's no change to the effective configs on x86_64, i686, aarch64 and armv7l. So merged.

teto commented:
@dezgeg mind sharing these scripts ? I want to do the same in https://github.com/NixOS/nixpkgs/pull/42838
